website/static/js/contribManager.js

```diff
@@ -6,106 +6,145 @@ var bootbox = require('bootbox');
 var $osf = require('./osfHelpers');
+require('js/filters');
 
-var contribsEqual = function(a, b) {
-    return a.id === b.id &&
-        a.visible === b.visible &&
-        a.permission === b.permission &&
-        Boolean(a.deleteStaged) === Boolean(b.deleteStaged);
-};
-
-// Modified from http://stackoverflow.com/questions/7837456/comparing-two-arrays-in-javascript
-var arraysEqual = function(a, b) {
-    var i = a.length;
-    if (i !== b.length) { return false;}
-    while (i--) {
-        if (!contribsEqual(a[i], b[i])) {return false;}
-    }
-    return true;
+//http://stackoverflow.com/questions/12822954/get-previous-value-of-an-observable-in-subscribe-of-same-observable
+ko.subscribable.fn.subscribeChanged = function (callback) {
+    var self = this;
+    var savedValue = self.peek();
+    return self.subscribe(function (latestValue) {
+        var oldValue = savedValue;
+        savedValue = latestValue;
+        callback(latestValue, oldValue);
+    });
 };
 
-var sortMap = {
-    surname: {
-        label: 'Surname',
-        order: 1,
-        func: function(item) {
-            return item.surname;
-        }
+ko.bindingHandlers.filters = {
+    init: function(element, valueAccessor, allBindingsAccessor, data, context) {
+        var $element = $(element);
+        var value = ko.utils.unwrapObservable(valueAccessor()) || {};
+        value.callback = data.callback;
+        $element.filters(value);
     }
 };
 
 // TODO: We shouldn't need both pageOwner (the current user) and currentUserCanEdit. Separate
 // out the permissions-related functions and remove currentUserCanEdit.
-var ContributorModel = function(contributor, currentUserCanEdit, pageOwner, isRegistration, isAdmin) {
+var ContributorModel = function(contributor, currentUserCanEdit, pageOwner, isRegistration, isAdmin, index, parent) {
 
     var self = this;
     $.extend(self, contributor);
 
-    self.permissionList = [
-        {value: 'read', text: 'Read'},
-        {value: 'write', text: 'Read + Write'},
-        {value: 'admin', text: 'Administrator'}
-    ];
-    self.getPermission = function(permission) {
-        for(var i = 0; i < self.permissionList.length; i++) {
-            if(permission === self.permissionList[i].value) {
-                return self.permissionList[i];
+    self.originals = {
+        permission: contributor.permission,
+        visible: contributor.visible,
+        index: index
+    };
+
+    self.toggleExpand = function() {
+        self.expanded(!self.expanded());
+    };
+
+    self.expanded = ko.observable(false);
+
+    self.filtered = ko.observable(false);
+
+    self.permission = ko.observable(contributor.permission);
+
+    self.permissionText = ko.observable(parent.permissionDict[self.permission()]);
+
+    self.visible = ko.observable(contributor.visible);
+
+    self.permission.subscribeChanged(function(newValue, oldValue) {
+        if (oldValue === 'admin') {
+            parent.adminCount(parent.adminCount() - 1);
```

Comment:
I think this is fine for now, but it would be preferable if `ContributorModel` didn't have to know so much about the `ContributorList`'s data members.

```js
function ContribList() {

    // for each contrib...
       var contrib = new ContribModel(... {onPermissionChanged: self.handlePermissionChanged})

    self.handlePermissionChanged = function(contrib, newPerm, oldPerm) {
        if (oldPerm === 'admin') {
            self.adminCount(parent.adminCount() - 1);
        }
    }
}

function ContribModel(..., opts) {
    //
    self.permission.subscribeChanged(function(newValue, oldValue) {
        self.options.onPermissionChanged(self, newValue, oldValue);
    }
}
```

This is better encapsulation, and makes for simpler reusability and testing.

That said, this isn't a blocker.